proposals/record-structs.md

```diff
@@ -0,0 +1,303 @@
+
+The record struct implements `System.IEquatable<R>` and includes a synthesized strongly-typed overload of `Equals(R other)` where `R` is the record struct.
+The method is `public`.
+The method can be declared explicitly. It is an error if the explicit declaration does not match the expected signature or accessibility.
+
+If `Equals(R other)` is user-defined (not synthesized) but `GetHashCode` is not, a warning is produced.
+
+```C#
+public bool Equals(R other);
+```
+
+The synthesized `Equals(R)` returns `true` if and only if for each instance field `fieldN` in the record struct
+the value of `System.Collections.Generic.EqualityComparer<TN>.Default.Equals(fieldN, other.fieldN)` where `TN` is the field type is `true`.
+
+The record struct includes synthesized `==` and `!=` operators equivalent to operators declared as follows:
+```C#
+public static bool operator==(R r1, R r2)
+    => r1.Equals(r2);
+public static bool operator!=(R r1, R r2)
+    => !(r1 == r2);
+```
+The `Equals` method called by the `==` operator is the `Equals(R other)` method specified above. The `!=` operator delegates to the `==` operator. It is an error if the operators are declared explicitly.
+
+The record struct includes a synthesized override equivalent to a method declared as follows:
+```C#
+public override bool Equals(object? obj);
+```
+It is an error if the override is declared explicitly. 
+The synthesized override returns `other is R && Equals((R)other)` where `R` is the record struct.
+
+The record struct includes a synthesized override equivalent to a method declared as follows:
+```C#
+public override int GetHashCode();
+```
+The method can be declared explicitly.
+
+A warning is reported if one of `Equals(R)` and `GetHashCode()` is explicitly declared but the other method is not explicit.
+
+The synthesized override of `GetHashCode()` returns an `int` result of a deterministic function combining the values of `System.Collections.Generic.EqualityComparer<TN>.Default.GetHashCode(fieldN)` for each instance field `fieldN` with `TN` being the type of `fieldN`.
+
+For example, consider the following record struct:
+```C#
+record struct R1(T1 P1, T2 P2);
+```
+
+For this record struct, the synthesized equality members would be something like:
+```C#
+struct R1 : IEquatable<R1>
+{
+    public T1 P1 { get; set; }
+    public T1 P1 { get; set; }
+    public override bool Equals(object? obj) => obj is R1 && Equals((R1)obj);
+    public virtual bool Equals(R1 other)
+    {
+        return
+            EqualityComparer<T1>.Default.Equals(P1, other.P1) &&
+            EqualityComparer<T2>.Default.Equals(P2, other.P2);
+    }
+    public static bool operator==(R1 r1, R1 r2)
+        => r1.Equals(r2);
+    public static bool operator!=(R1 r1, R1 r2)
+        => !(r1 == r2);    
+    public override int GetHashCode()
+    {
+        return Combine(
+            EqualityComparer<T1>.Default.GetHashCode(P1),
+            EqualityComparer<T2>.Default.GetHashCode(P2));
```

Comment:
should we mention that 'Combine' is not defined in the spec, but just has to be deterministic within an invocation of the program?